File: be_camphub/src/main/java/org/camphub/be_camphub/mapper/AccountMapper.java
package org.camphub.be_camphub.mapper;

import org.camphub.be_camphub.dto.request.account.AccountCreationRequest;
import org.camphub.be_camphub.dto.request.account.AccountPatchRequest;
import org.camphub.be_camphub.dto.request.account.AccountUpdateRequest;
import org.camphub.be_camphub.dto.request.auth.RegisterRequest;
import org.camphub.be_camphub.dto.response.account.AccountResponse;
import org.camphub.be_camphub.entity.Account;
import org.mapstruct.*;

@Mapper(componentModel = "spring", uses = Utils_Mapper.class)
public interface AccountMapper {

    @Mapping(target = "id", ignore = true)
    @Mapping(target = "trustScore", constant = "100")
    @Mapping(target = "coinBalance", constant = "0.0")
    @Mapping(target = "userType", constant = "USER")
    @Mapping(target = "status", constant = "ACTIVE")
    @Mapping(target = "createdAt", expression = "java(java.time.LocalDateTime.now())")
    @Mapping(target = "updatedAt", ignore = true)
    @Mapping(target = "deletedAt", ignore = true)
    Account registerRequestToEntity(RegisterRequest request);

    @Mapping(target = "id", ignore = true)
    @Mapping(target = "trustScore", constant = "100")
    @Mapping(target = "coinBalance", constant = "0.0")
    @Mapping(target = "userType", ignore = true)
    @Mapping(target = "status", constant = "ACTIVE")
    @Mapping(target = "avatar", ignore = true)
    @Mapping(target = "createdAt", expression = "java(java.time.LocalDateTime.now())")
    @Mapping(target = "updatedAt", ignore = true)
    @Mapping(target = "deletedAt", ignore = true)
    Account creationRequestToEntity(AccountCreationRequest request);

    @Mapping(target = "id", ignore = true)
    @Mapping(target = "username", ignore = true)
    @Mapping(target = "createdAt", ignore = true)
    @Mapping(target = "deletedAt", ignore = true)
    @Mapping(target = "updatedAt", expression = "java(java.time.LocalDateTime.now())")
    void updateRequestToEntity(@MappingTarget Account account, AccountUpdateRequest request);

    @BeanMapping(nullValuePropertyMappingStrategy = NullValuePropertyMappingStrategy.IGNORE)
    @Mapping(target = "id", ignore = true)
    @Mapping(target = "username", ignore = true)
    @Mapping(target = "createdAt", ignore = true)
    @Mapping(target = "deletedAt", ignore = true)
    @Mapping(target = "updatedAt", expression = "java(java.time.LocalDateTime.now())")
    void patchRequestToEntity(@MappingTarget Account account, AccountPatchRequest request);

    AccountResponse entityToResponse(Account account);
}

File: fe_camphub/src/libs/core/mapping/account.mapper.ts
// libs/mappers/account.mapper.ts
import { createMapper, createMap } from "@automapper/core";
import { pojos, PojosMetadataMap } from "@automapper/pojos";
import { Account } from "../types/account.model";
import { AccountResponse } from "../dto/response";
import {
    AccountCreationRequest,
    AccountUpdateRequest,
    AccountPatchRequest,
} from "../dto/request";

export const accountMapper = createMapper({
    strategyInitializer: pojos(),
});

PojosMetadataMap.create<Account>("Account", {
    id: String,
    username: String,
    firstname: String,
    lastname: String,
    email: String,
    phoneNumber: String,
    idNumber: String,
    avatar: String,
    trustScore: Number,
    coinBalance: Number,
    userType: String,
    status: String,
    createdAt: String,
    updatedAt: String,
});

PojosMetadataMap.create<AccountResponse>("AccountResponse", {
    id: String,
    username: String,
    firstname: String,
    lastname: String,
    email: String,
    phoneNumber: String,
    idNumber: String,
    avatar: String,
    trustScore: Number,
    coinBalance: Number,
    userType: String,
    status: String,
    createdAt: String,
    updatedAt: String,
});

PojosMetadataMap.create<AccountCreationRequest>("AccountCreationRequest", {
    username: String,
    password: String,
    firstname: String,
    lastname: String,
    email: String,
    phoneNumber: String,
    idNumber: String,
});

PojosMetadataMap.create<AccountUpdateRequest>("AccountUpdateRequest", {
    password: String,
    firstname: String,
    lastname: String,
    email: String,
    phoneNumber: String,
    idNumber: String,
    avatar: String,
    trustScore: Number,
    coinBalance: Number,
    userType: String,
    status: String,
});

PojosMetadataMap.create<AccountPatchRequest>("AccountPatchRequest", {
    password: String,
    firstname: String,
    lastname: String,
    email: String,
    phoneNumber: String,
    idNumber: String,
    avatar: String,
    trustScore: Number,
    coinBalance: Number,
    userType: String,
    status: String,
});

// Mapping
createMap<AccountResponse, Account>(accountMapper, "AccountResponse", "Account");
createMap<Account, AccountCreationRequest>(accountMapper, "Account", "AccountCreationRequest");
createMap<Account, AccountUpdateRequest>(accountMapper, "Account", "AccountUpdateRequest");
createMap<Account, AccountPatchRequest>(accountMapper, "Account", "AccountPatchRequest");

// Export
export const accountMap = {
    fromResponse: (dto: AccountResponse): Account =>
        accountMapper.map<AccountResponse, Account>(dto, "AccountResponse", "Account"),

    toCreationRequest: (model: Account): AccountCreationRequest =>
        accountMapper.map<Account, AccountCreationRequest>(model, "Account", "AccountCreationRequest"),

    toUpdateRequest: (model: Account): AccountUpdateRequest =>
        accountMapper.map<Account, AccountUpdateRequest>(model, "Account", "AccountUpdateRequest"),

    toPatchRequest: (model: Partial<Account>): AccountPatchRequest =>
        accountMapper.map<Partial<Account>, AccountPatchRequest>(model, "Account", "AccountPatchRequest"),
};